sklearn/utils/stats.py

```diff
@@ -29,41 +28,60 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     percentile : int if `array` 1D, ndarray if `array` 2D
         Weighted percentile.
     """
+    xp, _ = get_namespace(array)
     n_dim = array.ndim
     if n_dim == 0:
         return array[()]
     if array.ndim == 1:
-        array = array.reshape((-1, 1))
+        array = xp.reshape(array, (-1, 1))
     # When sample_weight 1D, repeat for each array.shape[1]
     if array.shape != sample_weight.shape and array.shape[0] == sample_weight.shape[0]:
-        sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
-    sorted_idx = np.argsort(array, axis=0)
-    sorted_weights = np.take_along_axis(sample_weight, sorted_idx, axis=0)
+        sample_weight = xp.tile(sample_weight, (array.shape[1], 1)).T
+    sorted_idx = xp.argsort(array, axis=0)
+    sorted_weights = _take_along_axis(sample_weight, sorted_idx, xp)
 
     # Find index of median prediction for each sample
     weight_cdf = stable_cumsum(sorted_weights, axis=0)
```

Comment:
Actually I think we need to convert `cumsum` into array API compatible, because when the device and namespace are not numpy (like torch) we get the numpy not available error.

Reply:
I think scikit-learn's `stable_cumsum` is overly complicated, and we could just directly call `np.cumsum` in the numpy case.

For array API compatibility, https://github.com/data-apis/array-api/pull/653 has been merged and that means that it will become part of the standard. We can probably anticipate this and have something like:

```python
def _cumsum(a, axis=None, dtype=None, xp=None):
    xp, _ = get_namespace(X, xp=xp)
    if hasattr(xp, "cumsum"):
        return xp.cumsum(a, axis=axis, dtype=dtype)

    return xp.asarray(np.cumsum(_convert_to_numpy(a), axis=axis), dtype=dtype)
```

and hope that `array_api_compat` will be updated quickly once the new spec is in published.

Reply:
Thank you for the suggested changes!! I am running the test locally and I keep on getting this error ```IndexError: self.ndim=2, but the multi-axes index only specifies 1 dimensions. If this was intentional, add a trailing ellipsis (...) which expands into as many slices (:) as necessary - this is what np.ndarray arrays implicitly do, but such flat indexing behaviour is not specified in the Array API.``` on the line after this one. Looking at the verbose outputs, I am guessing it's because `cumsum`'s output is a 2D array that contains only 1D data (ie ```Array([[  1.],
       [  2.],
       [  3.],
       [  4.],
       [  5.],
       [  6.],
       [  7.],
       [  8.]...    [ 97.],
       [ 98.],
       [ 99.],
       [100.],
       [101.],
       [101.]], dtype=array_api_strict.float64)```
and the Array API's `_validate_index` functions does like that. 

I am wondering what is the best way around this? I think that if we can guarantee the output of `cumsum` to always contain 1D data, we could squeeze it before using an index on it.